store/sqlstore/retention_policy_store.go

```diff
@@ -0,0 +1,521 @@
+	s.CreateCompositeIndexIfNotExists("IDX_RetentionPolicies_DisplayName_Id", "RetentionPolicies",
+		[]string{"DisplayName", "Id"})
+	s.CreateIndexIfNotExists("IDX_RetentionPoliciesChannels_PolicyId", "RetentionPoliciesChannels", "PolicyId")
+	s.CreateIndexIfNotExists("IDX_RetentionPoliciesTeams_PolicyId", "RetentionPoliciesTeams", "PolicyId")
+	s.CreateCheckConstraintIfNotExists("RetentionPolicies", "PostDuration", "PostDuration > 0")
+	s.CreateForeignKeyIfNotExists("RetentionPoliciesChannels", "PolicyId", "RetentionPolicies", "Id", true)
+	s.CreateForeignKeyIfNotExists("RetentionPoliciesTeams", "PolicyId", "RetentionPolicies", "Id", true)
+}
+
+// executePossiblyEmptyQuery only executes the query if it is non-empty. This helps avoid
+// having to check for MySQL, which, unlike Postgres, does not allow empty queries.
+func executePossiblyEmptyQuery(txn *gorp.Transaction, query string, args ...interface{}) (sql.Result, error) {
+	if query == "" {
+		return nil, nil
+	}
+	return txn.Exec(query, args...)
+}
+
+func (s *SqlRetentionPolicyStore) Save(policy *model.RetentionPolicyWithTeamAndChannelIds) (*model.RetentionPolicyWithTeamAndChannelCounts, error) {
+	// Strategy:
+	// 1. Insert new policy
+	// 2. Insert new channels into policy
+	// 3. Insert new teams into policy
+	var (
+		rpInsertQuery  string
+		rpInsertArgs   []interface{}
+		rpcInsertQuery string
+		rpcInsertArgs  []interface{}
+		rptInsertQuery string
+		rptInsertArgs  []interface{}
+		rpSelectQuery  string
+		rpSelectProps  map[string]interface{}
+	)
+
+	if err := s.checkTeamsExist(policy.TeamIds); err != nil {
+		return nil, err
+	}
+	if err := s.checkChannelsExist(policy.ChannelIds); err != nil {
+		return nil, err
+	}
+
+	policy.Id = model.NewId()
+
+	rpInsertQuery, rpInsertArgs, _ = s.getQueryBuilder().
+		Insert("RetentionPolicies").
+		Columns("Id", "DisplayName", "PostDuration").
+		Values(policy.Id, policy.DisplayName, policy.PostDuration).
+		ToSql()
+
+	rpcInsertQuery, rpcInsertArgs = s.buildInsertRetentionPoliciesChannelsQuery(policy.Id, policy.ChannelIds)
+
+	rptInsertQuery, rptInsertArgs = s.buildInsertRetentionPoliciesTeamsQuery(policy.Id, policy.TeamIds)
+
+	rpSelectQuery, rpSelectProps = s.buildGetPolicyWithCountsQuery(policy.Id)
+
+	txn, err := s.GetMaster().Begin()
+	if err != nil {
+		return nil, err
+	}
+	defer finalizeTransaction(txn)
+	if _, err = txn.Exec(rpInsertQuery, rpInsertArgs...); err != nil {
+		return nil, err
+	}
+	if _, err = executePossiblyEmptyQuery(txn, rpcInsertQuery, rpcInsertArgs...); err != nil {
+		return nil, err
+	}
+	if _, err = executePossiblyEmptyQuery(txn, rptInsertQuery, rptInsertArgs...); err != nil {
+		return nil, err
+	}
+	var newPolicy model.RetentionPolicyWithTeamAndChannelCounts
+	if err = txn.SelectOne(&newPolicy, rpSelectQuery, rpSelectProps); err != nil {
+		return nil, err
+	}
+	if err = txn.Commit(); err != nil {
+		return nil, err
+	}
+	return &newPolicy, nil
+}
```

Comment:
This method would be cleaner if you created the queries, just before running them. 
```
rpcInsertQuery, rpcInsertArgs = s.buildInsertRetentionPoliciesChannelsQuery(policy.Id, policy.ChannelIds)
if _, err = executePossiblyEmptyQuery(txn, rpcInsertQuery, rpcInsertArgs...); err != nil {
	return nil, err
}
```
As it is, I find it hard to follow.

Reply:
The reason why I built the queries before starting the transaction is so that we wouldn't have to build the queries while the transaction is running. This reduces the amount of time for which the transaction is open.
Your point about the code being hard to follow is valid though - I'll try to add more comments and meaningful variable names to make it easier.